File: fsharp/kindergarten-garden/KindergartenGarden.fs
module KindergartenGarden

type Plant = 
    | Violets
    | Radishes
    | Clover
    | Grass

let charToPlant (c : char) : Plant =
    match c with
    | 'V' -> Plant.Violets
    | 'R' -> Plant.Radishes
    | 'C' -> Plant.Clover
    | 'G' -> Plant.Grass
    |  _  -> raise <| System.ArgumentOutOfRangeException()

let plants (diagram : string) (student : string) : Plant list = 
    let rows = diagram.Split([|'\n'|], System.StringSplitOptions.RemoveEmptyEntries)
    let idx = int student.[0] - 65
    [charToPlant rows.[0].[2*idx]; charToPlant rows.[0].[2*idx + 1]; charToPlant rows.[1].[2*idx]; charToPlant rows.[1].[2*idx + 1]]

File: csharp/kindergarten-garden/KindergartenGarden.cs
using System;
using System.Collections.Generic;

public enum Plant
{
    Violets,
    Radishes,
    Clover,
    Grass
}

public class KindergartenGarden
{
    private string _firstRow;
    private string _secondRow;
    public KindergartenGarden(string diagram)
    {
        string[] rows = diagram.Split("\n");
        _firstRow = rows[0];
        _secondRow = rows[1];
    }

    public IEnumerable<Plant> Plants(string student)
    {
        int idx = (int)student[0] - 65;
        yield return charToPlant(_firstRow[2 * idx]);
        yield return charToPlant(_firstRow[2 * idx + 1]);
        yield return charToPlant(_secondRow[2 * idx]);
        yield return charToPlant(_secondRow[2 * idx + 1]);
    }

    private Plant charToPlant(char c)
    {
        return c switch
        {
            'V' => Plant.Violets,
            'R' => Plant.Radishes,
            'C' => Plant.Clover,
            'G' => Plant.Grass,
            _ => throw new ArgumentOutOfRangeException()
        };
    }
}